# Homework 1 Code and Report
AUFAR Rezka and LASBLEIS Alexandre 

### Step 1 : Simulate Rankings of Relevance for E and P (5 points)

In the first step you will generate pair s of rankings of relevance, for the production P and experimental E,
respectively, for a hypothetical query q . Assume a 3-graded relevance, i.e. {N, R, HR}. Construct all
possible P and E ranking pairs of length 5, for which E outperforms P.

Example:<br>
P: {N N N N N}<br>
E: {N N N N R}<br>
… <br>
P: {HR HR HR HR R}<br>
E: {HR HR HR HR HR}<br>

(Note 1: If you do not have enough computational power, sample 1000 pair uniformly at random to show
your work.)
##### Code

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import random as rand
from scipy.stats import bernoulli

In [2]:
#convert int to relevance
#_list : list of integer to be converted
#return : list of string where every integer has been replaced by its relevance interpretation
def to_string(_list): 
    L=[]
    for i in _list:
        if i==0:
            L.append("N ")
        elif i==1:
            L.append("R ")
        elif i==2:
            L.append("HR")
        else: 
            L.append("? ")
    return L

In [3]:
pairs = [] #variable used to store all pairs 
Ps = []
Es = []

#Create all combinaison for P, there are 3^5 = 243 possibilities since P is of length 5 and each member have 3 possible values:
# 0 (representing "N" ), 1 (representing "R), 2 (representing "HR)
for p_possibility in range(243): 
    P=[None]*5
    #create all five values for P 
    for p_pos in range(5):
        P[p_pos]= p_possibility // (np.power(3,p_pos)) %3 #formula to create all combinaison
    #we do the same for E    
    for e_possibility in range(243): 
        E=[None]*5
        for e_pos in range(5):
            E[e_pos]= e_possibility // (np.power(3,e_pos)) %3 #formula to create all combinaison
        pairs.append((P,E)) #adding the pair since every pair is valid
        #Memorize all created Es but only once
        if p_possibility==0:
            Es.append(E)
        #memorize all created Ps
    Ps.append(P)

In [4]:
#TESTING
print(len(pairs)) #= 59049 which is equal to the expected length = 3 ^10 = 59049
#print(pairs) #print all pairs but this takes a lot of place

#print a random pair to show that it makes sense 
r = rand.randint(0,59049-1)
print("P: " , to_string(pairs[r][0]), "E: ",to_string(pairs[r][1]))

#check if Es is as expected 
print(len(Es))#expect 243 = 3^5 
print(Es)

59049
('P: ', ['N ', 'HR', 'N ', 'N ', 'HR'], 'E: ', ['N ', 'R ', 'N ', 'R ', 'N '])
243
[[0, 0, 0, 0, 0], [1, 0, 0, 0, 0], [2, 0, 0, 0, 0], [0, 1, 0, 0, 0], [1, 1, 0, 0, 0], [2, 1, 0, 0, 0], [0, 2, 0, 0, 0], [1, 2, 0, 0, 0], [2, 2, 0, 0, 0], [0, 0, 1, 0, 0], [1, 0, 1, 0, 0], [2, 0, 1, 0, 0], [0, 1, 1, 0, 0], [1, 1, 1, 0, 0], [2, 1, 1, 0, 0], [0, 2, 1, 0, 0], [1, 2, 1, 0, 0], [2, 2, 1, 0, 0], [0, 0, 2, 0, 0], [1, 0, 2, 0, 0], [2, 0, 2, 0, 0], [0, 1, 2, 0, 0], [1, 1, 2, 0, 0], [2, 1, 2, 0, 0], [0, 2, 2, 0, 0], [1, 2, 2, 0, 0], [2, 2, 2, 0, 0], [0, 0, 0, 1, 0], [1, 0, 0, 1, 0], [2, 0, 0, 1, 0], [0, 1, 0, 1, 0], [1, 1, 0, 1, 0], [2, 1, 0, 1, 0], [0, 2, 0, 1, 0], [1, 2, 0, 1, 0], [2, 2, 0, 1, 0], [0, 0, 1, 1, 0], [1, 0, 1, 1, 0], [2, 0, 1, 1, 0], [0, 1, 1, 1, 0], [1, 1, 1, 1, 0], [2, 1, 1, 1, 0], [0, 2, 1, 1, 0], [1, 2, 1, 1, 0], [2, 2, 1, 1, 0], [0, 0, 2, 1, 0], [1, 0, 2, 1, 0], [2, 0, 2, 1, 0], [0, 1, 2, 1, 0], [1, 1, 2, 1, 0], [2, 1, 2, 1, 0], [0, 2, 2, 1, 0], [1, 2, 2, 1, 0], [2, 2, 2,

##### Explanations
In order to contruct all possible pairs of P and E with a 3-graded relevance {N, R, HR}, we first encode those relevance as integers: <br>
* 0 for N <br>
* 1 for R <br>
* 2 for HR <br>
This encoding allows us to do easy comparison between elements.
Then we created all pairs using this formula : $value = \frac{possibility}{3^{pos}} [3]$ with $possibility \in \{0,...,243\}$ and $ pos \in \{0,1,2,3,4\}$.

This give the following sequence depending on the position of the element in the list: <br>
* 0: $\{0,1,2,0,1,2,0,1,2,0,1,2...\}$
* 1: $\{0,0,0,1,1,1,2,2,2,0,0,0...\}$
* 2: $\{0,0,0,0,0,0,0,0,0,1,1,1,...\}$
* ... and so on

This formula is then used for the production P and the experiment E. 

##### Analysis
According to the formula we showed before, we should find all the possible combination. Futhermore, the number of pairs found is equal to the one we expected $3^{10}$ and the number of E and P are $3^{5}$ each which is also what is expected. 


### Step 2 : Implement Evaluation Measures (15 points)

Implement 1 binary and 2 multi-graded evaluation measures out of the 7 measures mentioned above.

(Note 2: Some of the aforementioned measures require the total number of relevant and highly relevant
documents in the entire collection – pay extra attention on how to find this)

##### code

In [5]:
#TODO: Remove for evaluation 

#Recall at rank k 
def Binary_measure_recal(test_set, r):
    num_test_rel = 0 
    total_num_rel=0
    for i in range(len(test_set)):
        if test_set[i]>0:
            if i<r:
                num_test_rel+=1
            total_num_rel+=1
    return float(num_test_rel) / total_num_rel

#Average precision at rank k 
def Binary_measure_avg_precision(test_set, r):
    num_rel=0
    sum_precision = 0 
    total_num_rel=0
    temp="("
    for i in range(len(test_set)):
        if test_set[i]>0:
            if i<r:
                num_rel+=1
                temp+=" "+str(num_rel)+"/"+str(i+1)+"+"
                sum_precision+=float(num_rel)/(i+1)
            total_num_rel+=1
    temp=temp[:len(temp)-1]+")"
    #print(temp,"/",total_num_rel)
    return float(sum_precision) / total_num_rel
#Normalized discount cumulative gain nDCG at rank r
def nDCG(test_set, r):
    orderedSet= list(test_set)
    orderedSet.sort(reverse=True)
    return DCG(test_set, r) / DCG(orderedSet, r)

In [6]:
#BINARY 
#Precision at rank r 
#test_set: the collection to be tested 
#r: the rank for which we test the collection 
#return : the precision 
def Binary_measure_precision(test_set, r):
    assert r <= len(test_set)
    assert r!= 0
    num_test_rel = 0 
    for i in range(r):
        #According to our notation, if test_set[i]>0, the document is relevant
        if test_set[i]>0:
            num_test_rel+=1
    return float(num_test_rel) / r #precision is number of relevant doc/ number of doc considered

#MULTI GRADED 
#Discount cumulative gain DCG at rank r
#test_set: the collection to be tested 
#r: the rank for which we test the collection 
#return : the gain 
def DCG(test_set, r):
    assert r <= len(test_set)
    assert r!= 0
    res=0
    for k in range(1,r+1):#r is the rank so it starts at 1
        rel_r = test_set[k-1]#index of element = rank -1 since rank starts at 1 and index at 0
        res+=float(np.power(2,rel_r)-1)/(np.log2(1+k))
    return res

#Rank Biased Precision w ith persistence parameter 𝜃 =0.8
#test_set: the collection to be tested 
#theta: the theta parameter of the RBP formula
#return : the precision
def RBP (test_set,theta=0.8):
    res=0
    for k in range(1,len(test_set)+1):
        rel_k=test_set[k-1]
        res+= rel_k * np.power(theta,k-1) * (1.0-theta)
    return res
        

In [7]:
#TESTING
test=        [1,0,2,5,0,1,0,0,0,1]#length 10
test_perfect=[5,2,1,1,1,0,0,0,0,0]
test_worst=  [0,0,0,0,0,1,1,1,2,5]

#The following results are rounded in order to make them readable
#Specific ordering 
print("Specific ordering")
for r in range(1,len(test)+1):
    print('Rank = ' , r,'Precision',round(Binary_measure_precision(test,r),3),'DCG', round(DCG(test,r),3),
          'RBP', round(RBP(test),3))

#Perfect ordering 
print("\nPerfect ordering")
for r in range(1,len(test_perfect)+1):
    print('Rank = ' , r,'Precision',round(Binary_measure_precision(test_perfect,r),3),'DCG', round(DCG(test_perfect,r),3),
          'RBP', round(RBP(test_perfect),3))

#Worst ordering 
print("\nWorst ordering")
for r in range(1,len(test_worst)+1):
    print('Rank = ' , r,'Precision',round(Binary_measure_precision(test_worst,r),3),'DCG', round(DCG(test_worst,r),3),
          'RBP', round(RBP(test_worst),3))

Specific ordering
('Rank = ', 1, 'Precision', 1.0, 'DCG', 1.0, 'RBP', 1.06)
('Rank = ', 2, 'Precision', 0.5, 'DCG', 1.0, 'RBP', 1.06)
('Rank = ', 3, 'Precision', 0.667, 'DCG', 2.5, 'RBP', 1.06)
('Rank = ', 4, 'Precision', 0.75, 'DCG', 15.851, 'RBP', 1.06)
('Rank = ', 5, 'Precision', 0.6, 'DCG', 15.851, 'RBP', 1.06)
('Rank = ', 6, 'Precision', 0.667, 'DCG', 16.207, 'RBP', 1.06)
('Rank = ', 7, 'Precision', 0.571, 'DCG', 16.207, 'RBP', 1.06)
('Rank = ', 8, 'Precision', 0.5, 'DCG', 16.207, 'RBP', 1.06)
('Rank = ', 9, 'Precision', 0.444, 'DCG', 16.207, 'RBP', 1.06)
('Rank = ', 10, 'Precision', 0.5, 'DCG', 16.496, 'RBP', 1.06)

Perfect ordering
('Rank = ', 1, 'Precision', 1.0, 'DCG', 31.0, 'RBP', 1.632)
('Rank = ', 2, 'Precision', 1.0, 'DCG', 32.893, 'RBP', 1.632)
('Rank = ', 3, 'Precision', 1.0, 'DCG', 33.393, 'RBP', 1.632)
('Rank = ', 4, 'Precision', 1.0, 'DCG', 33.823, 'RBP', 1.632)
('Rank = ', 5, 'Precision', 1.0, 'DCG', 34.21, 'RBP', 1.632)
('Rank = ', 6, 'Precision', 0.833, 'DCG', 34.2

##### Explanations
For Precision and DCG, we need to ensure that rank k does not exceed the documents length, and k cannot be zero.<br>
For DCG and RBP, we also decided that the value of the relevant would be equal to the value at that rank in the list. For example if an example is extremely relevant, he might get a value of 5 (as in the example) which leads to higher gain in DCG than a document which is only highly relevant (value of 2).<br>
Otherwise our implementation follows the definition of the method. 

##### Analysis
The result for Precision is highly dependant on the number of non-relevant in the cut-off collection. As k increases, the more non-relevant exists in the cut-off, the lower the precision will be. If we consider a rank equal to the length of the list, the precision is the same whatever the order is. Precision does not that into account how good the relevance of a document is, it only cares if a document is relevant or not. 

For DCG, the result depends on the degree of the relevance and the order of it. If the highest relevant score exists in the first position of the cut-off, the higher DCG will be. We showed that by displaying the DCG score with perfect ordering which is higher at any rank than the "specific ordering". Those value may be hard to use to compare different collections due to the fact that it is sensitive to the degree of relevance. A way to cope with that problem is to use its alternative version: the nDCG (normalized version). 

For RBP, we disregard rank and therefore we have the same result no matter the rank is. This measure pays great attention to the ordering as shown above: the best score is obtained with a perfect ordering whereas the worst one is obtained with the worst ordering. The model behind RBP considers that when you see a document, you have a probability to see the next document otherwise you stop. This model can be improved by taking into account how relevant is the document that we are looking at. Depending on this relevance, there is a probability to see the next document or to stop. This second model is called ERR model.



### Step 3 : Calculate the measure (5 points)

For the three measures and all P and E ranking pairs constructed above calculate the difference: 𝛥measure
= measure E -measure P . Consider only those pairs for which E outperforms P.

##### code

In [8]:
#Evaluate if E outperfoms P with respect to a specific measure. 
#methond_name: the measure used to test. It can be "precision", "DCG" or "RBP"
#E: 
#P:
#rank: 
#return : (outperforms, delta measure) where outperforms is True if E outperfors P with respect to the measure method for the
#rank and delta measure is the difference between to measure of E and P for the measure method.
def outperforms(methond_name,E,P,rank):
    if methond_name=="precision":
        delta = Binary_measure_precision(E,rank)-Binary_measure_precision(P,rank)
        return (delta>0), delta 
    elif methond_name=="DCG":
        delta=DCG(E,rank)-DCG(P,rank)
        return (delta>0), delta 
    elif methond_name=="RBP":
        delta=RBP(E)-RBP(P)
        return (delta>0), delta 

#variable for precision measure 
outperf_pairs_prec=[] #memorize pairs for which E outperfoms P
pairs_set_prec = [] 
d_measure_prec=[] #memorize "measure(E)-measure(P)" for which E outperfoms P

#variable for DCG measure 
outperf_pairs_DCG=[] #memorize pairs for which E outperfoms P
pairs_set_DCG = []
d_measure_DCG=[] #memorize "measure(E)-measure(P)" for which E outperfoms P

#variable for RBP measure
outperf_pairs_RBP=[] #memorize pairs for which E outperfoms P
pairs_set_RBP = []
d_measure_RBP=[] #memorize "measure(E)-measure(P)" for which E outperfoms P

rank = 4
for i in range(len(pairs)):#pairs made of (P,E)
    #test if E outperforms P with respect to a specific method
    outperforms_prec,delta_prec= outperforms("precision",pairs[i][1],pairs[i][0],rank)
    outperforms_DCG,delta_DCG= outperforms("DCG",pairs[i][1],pairs[i][0],rank)
    outperforms_RBP,delta_RBP= outperforms("RBP",pairs[i][1],pairs[i][0],rank)

    #Memorize some "precision" information only if E outperfom P
    if(outperforms_prec):
        outperf_pairs_prec.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_prec.append(''.join(str(x) for x in temp))
        d_measure_prec.append(delta_prec)
    
    #Memorize some "DCG" information only if E outperfom P
    if(outperforms_DCG):
        outperf_pairs_DCG.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_DCG.append(''.join(str(x) for x in temp))
        d_measure_DCG.append(delta_DCG)
    
    #Memorize some "RBP" information only if E outperfom P
    if(outperforms_RBP):
        outperf_pairs_RBP.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_RBP.append(''.join(str(x) for x in temp))
        d_measure_RBP.append(delta_RBP)     


In [11]:
#TESTING

print ("Number of pairs for which E outperforms P for all three measures: ",len(set.intersection(set(pairs_set_prec), set(pairs_set_DCG), set(pairs_set_RBP))))
print ("Pair in precision and in DCG:",rand.sample(set.intersection(set(pairs_set_prec), set(pairs_set_DCG)),1))

# pick result that is in prec but is not in DCG
print ("Pair in precision and not in DCG:",rand.sample(set(pairs_set_prec) - set(pairs_set_DCG),1))
# pick result that is in DCG but is not in prec
print ("Pair in DCG and not in precision:",rand.sample(set(pairs_set_DCG) - set(pairs_set_prec),1))


Number of pairs for which E outperforms P for all three measures:  17184
Pair in precision and in DCG: ['1102100122']
Pair in precision and not in DCG: ['1111222102']
Pair in DCG and not in precision: ['2222212111']


In [9]:
#TESTING
pair_prec_not_DCG = [[1,1,1,1,2],[2,2,1,0,2]] # pair E, P 
pair_DCG_not_prec = [[2,2,2,2,2],[1,2,1,1,1]] # pair E, P 
print("Pair in precision and not in DCG:",pair_prec_not_DCG, "precision E:",
      Binary_measure_precision(pair_prec_not_DCG[0],rank),"precision P:",
      Binary_measure_precision(pair_prec_not_DCG[1],rank),"DCG E: ",
     DCG(pair_prec_not_DCG[0],rank),"DCG P: ",
     DCG(pair_prec_not_DCG[1],rank) )

print("Pair in DCG and not in precision:",pair_DCG_not_prec, "DCG E: ",
     DCG(pair_DCG_not_prec[0],rank),"DCG P: ",
     DCG(pair_DCG_not_prec[1],rank),"precision E:",
      Binary_measure_precision(pair_DCG_not_prec[0],rank),"precision P:",
      Binary_measure_precision(pair_DCG_not_prec[1],rank))

('Pair in precision and not in DCG:', [[1, 1, 1, 1, 2], [2, 2, 1, 0, 2]], 'precision E:', 1.0, 'precision P:', 0.75, 'DCG E: ', 2.5616063116448506, 'DCG P: ', 5.3927892607143724)
('Pair in DCG and not in precision:', [[2, 2, 2, 2, 2], [1, 2, 1, 1, 1]], 'DCG E: ', 7.6848189349345519, 'DCG P: ', 3.8234658187877653, 'precision E:', 1.0, 'precision P:', 1.0)


##### Explanations
For precision, DCG, and RBP, we calculate the difference between all combination of E and P that we have generated in Step 1, then we append every difference that is greater than zero, which means that E outperforms P.
##### Analysis
For each three measures, we have different pairs in which E outperforms P. There are 17184 pairs where E outperforms P in all three measures. However, there exists some pairs where E outperforms P for one kind of measure, but not the other. <br> 
For example, E=[1,1,1,1,2] outperforms P=[2,2,1,0,2] in precision (delta of 0.25), but not in DCG (delta of -2.83118294971) . This is due to the fact that precision does not check order and the degree of relevance. The same happens the opposite way : E=[2,2,2,2,2] outperforms P=[1,2,1,1,1] in DCG (delta of 3.86135311614), but not in precision (delta of 0). <br>
This indicates that in real world application, it is important to implement different kinds of measure so we will have broader feedback to optimize the algorithm.

## Step 4 : Implement Interleave (15 points)

Implement 2 interleaving algorithms: (1) Team-Draft Interleaving OR Balanced Interleaving, AND (2),
Probabilistic Interleaving. The interleaving algorithms should (a) given two rankings of relevance
interleave them into a single ranking, and (b) given the users clicks on the interleaved ranking assign
credit to the algorithms that produced the rankings.

##### code

In [117]:
# assignment assumption : E and P always return different documents
# general parameter : consider two pairs can contain same documents 

def team_draft_interleave(ranking_pairs, general=False):
    # interleave two list
    interleaved_list = []
    relevance_list = []
    added_list = []
    assert len(ranking_pairs[0]) ==  len(ranking_pairs[1])
    for i, elem in enumerate(ranking_pairs[0]):
        coin = rand.choice([0,1])
        if(coin == 0):
            if(general and ranking_pairs[0][i] not in added_list):
                interleaved_list.append((ranking_pairs[0][i], "E"))
                added_list.append(ranking_pairs[0][i])
            if(general and ranking_pairs[1][i] not in added_list):
                interleaved_list.append((ranking_pairs[1][i], "P"))
                added_list.append(ranking_pairs[1][i])
            if(not general):
                interleaved_list.append((ranking_pairs[0][i], "E"))
                interleaved_list.append((ranking_pairs[1][i], "P"))
                relevance_list.append(ranking_pairs[0][i])
                relevance_list.append(ranking_pairs[1][i])
        if(coin == 1):
            if(general and ranking_pairs[1][i] not in added_list):
                interleaved_list.append((ranking_pairs[1][i], "P"))
                added_list.append(ranking_pairs[1][i])
            if(general and ranking_pairs[0][i] not in added_list):
                interleaved_list.append((ranking_pairs[0][i], "E"))
                added_list.append(ranking_pairs[0][i])
            if(not general):
                interleaved_list.append((ranking_pairs[1][i], "P"))
                interleaved_list.append((ranking_pairs[0][i], "E"))
                relevance_list.append(ranking_pairs[1][i])
                relevance_list.append(ranking_pairs[0][i])
    return interleaved_list, relevance_list

In [118]:
# TESTING
# tested with general case, which E and P can contain same link
general_test = ([1,2,3,4,5], [6,7,8,1,2])
print team_draft_interleave(general_test, True)

([(6, 'P'), (1, 'E'), (7, 'P'), (2, 'E'), (3, 'E'), (8, 'P'), (4, 'E'), (5, 'E')], [])


In [119]:
# TESTING
# tested the result with existing library (https://github.com/mpkato/interleaving/tree/master/interleaving)

import interleaving
a = [1, 2, 3, 4, 5] # Ranking 1
b = [6, 7, 8, 9, 10] # Ranking 2
method = interleaving.Probabilistic([a, b]) # initialize an interleaving method
ranking = method.interleave() # interleaving
print ranking

clicks = [0, 3] # observed clicks, i.e. documents 1 and 2 are clicked
result = interleaving.Probabilistic.evaluate(ranking, clicks)
print result # (0, 1) indicates Ranking 1 won Ranking 2.

[7, 1, 3, 2, 6]
[]


In [134]:
import copy

def softmax(tau, length):
    total_denominator = 0
    softmax_distribution = []
    for i in range(1, length+1):
        total_denominator += float(1.0/np.power(i,tau))
    for i in range(1, length+1):
        softmax_distribution.append(float(1.0/np.power(i,tau) / (total_denominator)))
    return softmax_distribution

def probabilistic_interleave(ranking_pairs):
    # generate softmax probability distribution with tau = 3 (Hofmann, 2011)
    assert len(ranking_pairs[0]) ==  len(ranking_pairs[1])
    E = copy.copy(ranking_pairs[0])
    P = copy.copy(ranking_pairs[1])
    softmax_distribution_E = softmax(3,len(E))
    softmax_distribution_P = softmax(3,len(P))
    # interleave two list with softmax probability distribution
    interleaved_list = []
    relevance_list = []
    for _ in range(len(E) + len(P)):
        coin = rand.choice([0,1])
        if(coin == 0):
            # check if E is empty then just sample from P
            if not E:
                # choose element from P randomly
                picked = np.random.choice(P, 1, p=softmax_distribution_P)[0]
                # append it to interleaved list
                interleaved_list.append((picked, "P"))
                relevance_list.append(picked)
                # remove the picked element
                P.remove(picked)
                # renormalize the distribution
                softmax_distribution_P = softmax(3,len(P))
            # else sample from E
            else:
                # choose element from E randomly
                picked = np.random.choice(E, 1, p=softmax_distribution_E)[0]
                # append it to interleaved list
                interleaved_list.append((picked, "E"))
                relevance_list.append(picked)
                # remove the picked element
                E.remove(picked)
                # renormalize the distribution
                softmax_distribution_E = softmax(3,len(E))
        if(coin == 1):
            # check if P is empty then just sample from E
            if not P:
                # choose element from E randomly
                picked = np.random.choice(E, 1, p=softmax_distribution_E)[0]
                # append it to interleaved list
                interleaved_list.append((picked, "E"))
                relevance_list.append(picked)
                # remove the picked element
                E.remove(picked)
                # renormalize the distribution
                softmax_distribution_E = softmax(3,len(E))
            # else sample from P
            else:
                # choose element from P randomly
                picked = np.random.choice(P, 1, p=softmax_distribution_P)[0]
                # append it to interleaved list
                interleaved_list.append((picked, "P"))
                relevance_list.append(picked)
                # remove the picked element
                P.remove(picked)
                # renormalize the distribution
                softmax_distribution_P = softmax(3,len(P))
    # generate user clicks randomly
    return interleaved_list, relevance_list

In [135]:
r = rand.randint(0,59049-1)

In [136]:
def who_wins(interleaved_list, clicks):
    # calculate scores
    E_wins, P_wins = 0, 0
    for i, click in enumerate(clicks):
        if click == 1:
            if interleaved_list[i][1] == "E":
                E_wins += 1
            if interleaved_list[i][1] == "P":
                P_wins += 1
    if E_wins > P_wins:
        return "E win"
    elif E_wins < P_wins:
        return "P Win"
    else:
        return "tie"

In [137]:
# TESTING

click_test = [1,1,0,0,1,0,0,0,0,0]
td_interleaved = team_draft_interleave(pairs[r], False)
prob_interleaved = probabilistic_interleave(pairs[r])
print pairs[r]
print td_interleaved
print prob_interleaved
print who_wins(td_interleaved[0], click_test)
print who_wins(prob_interleaved[0], click_test)

([0, 0, 2, 2, 1], [1, 1, 1, 2, 0])
([(1, 'P'), (0, 'E'), (1, 'P'), (0, 'E'), (2, 'E'), (1, 'P'), (2, 'E'), (2, 'P'), (1, 'E'), (0, 'P')], [1, 0, 1, 0, 2, 1, 2, 2, 1, 0])
([(1, 'P'), (0, 'E'), (1, 'P'), (0, 'E'), (1, 'P'), (2, 'P'), (2, 'E'), (2, 'E'), (0, 'P'), (1, 'E')], [1, 0, 1, 0, 1, 2, 2, 2, 0, 1])
E win
P Win


To do probabilistic interleave : 
- Make it general
- Explanation and analysis

##### Explanations


##### Analysis


## Step 5 : Implement User Clicks Simulation ( 25 points)

Having interleaved all the ranking pairs an online experiment could be ran. However, given that we do not
have any users (and the entire homework is a big simulation) we will simulate user clicks.<br>
We have considered a number of click models including:<br>
* 1. Random Click Model (RCM)
* 2. Position-Based Model (PBM)
* 3. Simple Dependent Click Model (SDCM)
* 4. Simple Dynamic Bayesian Network (SDBN)

Consider two different click models, (a) the Random Click Model (RCM), and (b) one out of the
remaining 3 aforementioned models. The parameters of some of these models can be estimated using the
Maximum Likelihood Estimation (MLE) method, while others require using the
Expectation-Maximization (EM) method. Implement the two models so that:
* (a) there is a method that learns the parameters of the model given a set of training data, 
* (b) there is a method that predicts the click probability given a ranked list of relevance labels, 
* (c) there is a method that decides - stochastically -whether a document is clicked based on these probabilities.<br>

Having implemented the two click models, estimate the model parameters using the Yandex Click Log

##### code

In [56]:
#Random Click Model (RCM) 
def learns_param_RCM(training_data):
    numb_doc = 0
    num_click=0
    for data in training_data:
        #Only if the action is a query, we add to the total number of doc the number of doc returned by the query
        if("Q" in data[2]):
            for j in range(5,len(data)):
                numb_doc+=1
        #If the action is a click, we add one to the total number of click 
        if("C" in data[2]):
            num_click+=1
    return float(num_click)/numb_doc

def predict_click_proba_RCM(training_data,relevance_labels) :
    rau = learns_param_RCM(training_data)
    #since the probability is rau for every document, we return a list of rau that has the length of the relevance labels list
    return [rau]*len(relevance_labels)

def is_doc_clicked_RCM(training_data,relevance_labels,document):
    proba_distrib = predict_click_proba_RCM(training_data,relevance_labels)
    return (bernoulli.rvs(proba_distrib[document])==1)

def generate_click_RCM(training_data, interleaved_list):
    click = []
    proba_distrib = predict_click_proba_RCM(training_data,relevance_labels)
    for i, elem in enumerate(proba_distrib):
        click.append(bernoulli.rvs(elem))
    return click

#Simple Dynamic Bayesian Network (SDBN)

#This means that the attractiveness of a document, given a query, is directly proportional to the number
#of times this document is clicked and inversely proportional to the number of times it is presented in
#response to the given query above or at the last-clicked rank.
def add1(key,dictionnary):
    if key not in dictionnary:
        dictionnary[key]=1
    else:
        dictionnary[key]+=1

def learns_param_SBDN(training_data):
    #the number of times document u is presented as response for the query q : used to calculate attrativness
    doc_presented_q_dict= {}
    #the number of times document u is clicked for the query q: used to calculate attrativness and satisfaction
    doc_clicked_q_dict= {}
    #the number of times document u is the last clicked for the query q: used to calculate satisfaction
    doc_last_clicked_q_dict= {}
    
    global_index = 0 #index used to browser training_data
    for session in range(int(training_data[len(training_data)-1][0])):
        #index used to browser all the documents in a session
        i=0 
        end_of_session=False
        #memorise the last query id
        last_query =-1
        #loop to browser the current session
        while(not end_of_session):
            #the session stop if the next index is not equal to the current session index or we reach the end of the list
            if not (global_index+i+1)>= len(training_data):
                end_of_session=(not int(training_data[global_index+i+1][0]) == session)
            else:
                end_of_session=True
                
            #memorize the current element for efficency reasons
            element = training_data[global_index+i]
            #memorize the last query
            if(element[2]=='Q'):
                last_query=int(element[3])
                #memorize that a document appears for a query
                for j in range(5,len(element)):
                    key= (int(element[j]),last_query)
                    add1(key,doc_presented_q_dict)
            #memorize the click
            if(element[2]=='C'):
                key= (int(element[3]),last_query)
                #If this is the last click of a session: memorize in the last click dictionnary 
                if end_of_session:
                    add1(key,doc_last_clicked_q_dict)
                #memorize in the click dictionnary 
                add1(key,doc_clicked_q_dict)
            #increment the index used to browser the session 
            i+=1
        #after the end of a session, skip to the next session
        global_index+=i
    
    attract= {}
    satisf = {}
    #calculate attractivness which is #u is clicked for query q / #u appears with q 
    for key in doc_presented_q_dict:
        if key in doc_clicked_q_dict:
            attract[key]= float(doc_clicked_q_dict[key]) /  doc_presented_q_dict[key]
        else: 
            attract[key]=0
    #calculate satisfaction which is #u is last clicked document for query q / #u is clicked for query q 
    for key in doc_clicked_q_dict:
        if key in doc_last_clicked_q_dict:
            satisf[key]= float(doc_last_clicked_q_dict[key]) /  doc_clicked_q_dict[key]
        else: 
            satisf[key]=0
    return attract,satisf

#CHECK WITH REZKA: what about the conditionnal probability of a click ?? 
#query is the query number 
def predict_click_proba_SBDN(training_data,relevance_labels,doc_ids,query) :
    attract,satisf= learns_param_SBDN(training_data)
    proba= [0]* len(relevance_labels)
    exam_proba=[0]*len(relevance_labels)
    
    for i in len(doc_ids):
        alpha_i_q = 0
        key = (doc_ids[i],query)
        if key in attract:
            alpha_i_q = attract[key]
            
        sigma_i_q = 0 
        if key in satisf:
            sigma_i_q = satisf[key]
        
        if(i==0):
            proba[i]=alpha_i_q
            exam_proba[i]=1
        else :
            proba[i]=alpha_i_q * exam_proba[i-1]
            #gamma=1
            exam_proba[i]=exam_proba[i-1] * ( (1-sigma_i_q)*alpha_i_q + (1-alpha_i_q) )
    return proba

#docuement: document position in relevance_labels list 
def is_doc_clicked_SBDN(training_data,relevance_labels,doc_ids,query,document):
    proba_distrib= predict_click_proba_SBDN(training_data,relevance_labels,doc_ids,query)
    return (bernoulli.rvs(proba_distrib[document])==1)


__CHECK WITH REZKA__

$P(C_u=1) = \alpha_{uq} \epsilon_{r_u} $

if $(P(C_u=1)$

$\epsilon_{r+1}= \epsilon_{r}\gamma  [(1-\sigma_{uq})\alpha_{uq} +(1-\alpha_{uq})] $

else if $ (P(C_u=1|C_{<r_u})$

$\epsilon_{r+1}= c_r^{(s)}\gamma (1-\sigma_{u_r q}) + (1-c_r^{(s)}) \frac{(1-\alpha_{u_r q}) \epsilon_r \gamma}{1-\alpha_{u_r q}\epsilon_r}  $


In [146]:
#open file and preprocess data
training_data=[]
relevance_labels= [2, 1, 2, 1, 1, 0, 2, 1, 2, 2]
document= 3

f= open("YandexRelPredChallenge.txt","r")
for line in f:
    training_data.append(line.split("\n")[0].split('\t'))
f.close()

print("isDocClicked", is_doc_clicked_RCM(training_data,relevance_labels,document))
print generate_click_RCM(training_data, relevance_labels)

#print("isDocClicked", is_doc_clicked_SBDN(training_data,relevance_labels,document))
#print(learns_param_SBDN(training_data))

('isDocClicked', False)
[0, 1, 0, 0, 0, 0, 1, 0, 0, 0]


##### Explanations

##### Analysis

## Step 6 : Simulate Interleaving Experiment ( 10 points)

Having implemented the click models, it is time to run the simulated experiment.<br>
For each of interleaved ranking run N simulations for each one of the click models implemented and
measure the proportion p of wins for E.<br>

(Note 7: Some of the models above include an attractiveness parameter $\alpha_{uq}$ . Use the relevance label to
assign this parameter by setting $\alpha_{uq}$ for a document u in the ranked list accordingly.)
##### code

In [138]:
#For our case, we have to redined the attractivness and thus the previous functions of SBDN:
#query value is a random number generated by generate_query
#doc_ids ...                   generated by generate_doc_ids
def specific_learns_param_SBDN(training_data,relevance_labels,doc_ids,query):
    attract,satisf= learns_param_SBDN(training_data)
    #the previous attractivness is incorrect, we calculate it using relevance labels: 
    #thanks to formula p47 of "Click Models for Web Search",for a query q 
    
    #relevance(u) = satisf(u) * attract(u) => attract(u) = relevance(u) / satisf(u)
    relev_attract=[0]*len(relevance_labels)
    for i in range(relevance_labels):
        satisf_u = 0
        key = (doc_ids[i],query)
        if key in satisf:
            satisf_u = satisf[u]
        relev_attract[i]= relevance_labels[i]/satisf_u
    return relev_attract,satisf

def generate_query(training_data):
    query= set()
    for data in training_data:
        if data[2]=="Q":
            query.add(int(data[3]))
    return rand.sample(query,1)
    
def generate_doc_ids(training_data,number_ids, number_ids_different):
    doc_id_set = set()
    res=[]
    for data in training_data:
        if data[2]=="Q":
            for j in range(5,len(data)):
                doc_id_set.add(int(data[j]))
   
    for i in range(number_ids):
        doc_id = rand.sample(doc_id_set,1)[0]
        valid_id = (doc_id not in number_ids_different) and (doc_id not in res)
        while (not valid_id):
            doc_id = rand.sample(doc_id_set,1)[0]
            valid_id = (doc_id not in number_ids_different) and (doc_id not in res)
            valid_id= valid_id 
        res.append(doc_id)
    return res 
        
#query value is a random number generated by generate_query
#doc_ids ...                   generated by generate_doc_ids
def specific_predict_click_proba_SBDN(training_data,relevance_labels):
    attract,satisf= specific_learns_param_SBDN(training_data,relevance_labels, query, doc_ids)
    
    proba= [0]* len(relevance_labels)
    exam_proba=[0]*len(relevance_labels)
    
    for i in len(relevance_labels):
        alpha_i_q = attract[i]
            
        sigma_i_q = 0 
        key = (doc_ids[i],query)
        if key in satisf [key]:
            sigma_i_q = satisf[key]
        
        if(i==0):
            proba[i]=alpha_i_q
            exam_proba[i]=1
        else :
            proba[i]=alpha_i_q * exam_proba[i-1]
            #gamma=1
            exam_proba[i]=exam_proba[i-1] * ( (1-sigma_i_q)*alpha_i_q + (1-alpha_i_q) )
    return proba
#query value is a random number generated by generate_query
#doc_ids ...                   generated by generate_doc_ids
def specific_is_doc_clicked_SBDN(training_data,relevance_labels,doc_ids,query,document):
    proba_distrib= specific_predict_click_proba_SBDN(training_data,relevance_labels,doc_ids,query)
    return (bernoulli.rvs(proba_distrib[document])==1)

In [155]:
test_pairs = rand.sample(pairs, 100)

In [156]:
E_wins_proportion_td = 0
E_wins_proportion_prob = 0
tie_td = 0
tie_prob = 0
for i, elem in enumerate(test_pairs):
    team_draft_res = team_draft_interleave(elem, False)
    prob_res = probabilistic_interleave(elem)
    clicks_td_RCM = generate_click_RCM(training_data, team_draft_res[1])
    clicks_prob_RCM = generate_click_RCM(training_data, prob_res[1])
    if(who_wins(team_draft_res[0], clicks_td_RCM) == "E win"):
        E_wins_proportion_td += 1
    elif(who_wins(team_draft_res[0], clicks_td_RCM) == "tie"):
        tie_td += 1
    if(who_wins(prob_res[0], clicks_prob_RCM) == "E win"):
        E_wins_proportion_prob += 1
    elif(who_wins(prob_res[0], clicks_prob_RCM) == "tie"):
        tie_prob += 1

print float(E_wins_proportion_td) / (100.0 - tie_td)
print float(E_wins_proportion_prob) / (100.0 - tie_prob)

0.460317460317
0.403846153846


To do :
- More verbality
- SDBN simulation
- Explanation and analysis